devito/function.py

```diff
@@ -1052,20 +1055,47 @@ def _is_owned(self, point):
         return all(distributor.glb_to_loc(d, p) is not None
                    for d, p in zip(self.grid.dimensions, point))
 
+    @property
+    def gridpoints(self):
+        """
+        The *reference* grid point corresponding to each sparse point.
+        """
+        raise NotImplementedError
+
+    @property
+    def _support(self):
+        """
+        The grid points surrounding each sparse point within the radius of self's
+        injection/interpolation operators.
+        """
+        ret = []
+        for i in self.gridpoints:
+            support = [range(max(0, j - self._radius + 1), min(M, j + self._radius + 1))
+                       for j, M in zip(i, self.grid.shape)]
+            ret.append(tuple(product(*support)))
+        return ret
+
     @property
     def _dist_datamap(self):
         """
-        Return a mapper ``M : MPI rank -> logically owned sparse data``.
+        Return a mapper ``M : MPI rank -> required sparse data``.
         """
-        targets = self.grid.distributor.glb_to_rank(self.gridpoints)
-        return as_mapper(range(self.npoint), lambda i: targets[i])
+        ret = {}
+        for i, s in enumerate(self._support):
+            # Sparse point `i` is "required" by the following ranks
+            for r in self.grid.distributor.glb_to_rank(s):
+                ret.setdefault(r, []).append(i)
+        return {k: filter_ordered(v) for k, v in ret.items()}
 
     @property
-    def _dist_mask(self):
+    def _dist_scatter_mask(self):
         """
-        Return a mask, suitable to index into ``self.data``, usable to
-        reorder the sparse data values so that adjacent values logically
-        belong to the same MPI rank.
+        Return a mask to index into ``self.data``, which creates a new
+        data array that logically contains N consecutive groups of sparse
+        data values, where N is the number of MPI ranks. The i-th group
+        contains the sparse data values accessible by the i-th MPI rank.
+        Thus, sparse data values along the boundary of two or more MPI
+        ranks are duplicated.
         """
         dmap = self._dist_datamap
         mask = np.array(flatten(dmap[i] for i in sorted(dmap)), dtype=int)
```

Comment:
Only `dmap[self.rank]` should be scattered, the `alltoall` will take care of the receive part